README
To build run:
$ sudo apt-get build-dep goget-ubuntu-touch

?!?! there is surely a better way 
$ debian/rules clean build


debian/changelog
goget-ubuntu-touch (0.35+ubports) xenial; urgency=medium

  * Imported to UBports

 -- UBports auto importer <infra@ubports.com>  Thu, 11 Jan 2018 10:13:54 +0000

goget-ubuntu-touch (0.34+16.10.20160805-0ubuntu1) yakkety; urgency=medium

  [ Andrea Bernabei ]
  * Default to ubuntu-touch/stable instead of ubuntu-touch/devel for
    emulator creation (LP: #1594306)

 -- Łukasz Zemczak <lukasz.zemczak@canonical.com>  Fri, 05 Aug 2016 13:21:10 +0000

goget-ubuntu-touch (0.34+16.10.20160722-0ubuntu1) yakkety; urgency=medium

  [ Simon Fels ]
  * Add bootstrap support for turbo/frieza/cooler devices

 -- Łukasz Zemczak <lukasz.zemczak@canonical.com>  Fri, 22 Jul 2016 16:02:30 +0000

goget-ubuntu-touch (0.34-0ubuntu1) xenial; urgency=medium

  [ Ondrej Kubik ]
  * touch: new adb support with key support. by w-ondra approved by mvo
    LP: #1316302

 -- Michael Vogt <michael.vogt@ubuntu.com>  Tue, 05 Apr 2016 12:21:07 +0200

goget-ubuntu-touch (0.33-0ubuntu3) xenial; urgency=medium

  * remove support for building ubuntu-core (snappy) images
    until the kernel/gadget snap specification is finalized

 -- Michael Vogt <michael.vogt@ubuntu.com>  Tue, 05 Apr 2016 09:46:47 +0200

goget-ubuntu-touch (0.33-0ubuntu2) xenial; urgency=medium

  * debian/control: Remove version Build-Depends for golang-snappy-dev. 

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Wed, 18 Nov 2015 09:32:39 -0300

goget-ubuntu-touch (0.33-0ubuntu1) xenial; urgency=medium

  * Exit 1 on parameter errors (LP: #1473333) 

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Tue, 17 Nov 2015 16:56:44 -0300

goget-ubuntu-touch (0.32-0ubuntu1) xenial; urgency=medium

  [ Michael Vogt ]
  * Increase the size of the boot partition to 128M
  * create fw/ directory under the efi directory
  * support 'dst' keyword in boot-assets
  * update build-dep version for golang-snappy-dev

  [ Sergio Schvezov ]
  * updated for upstream snappy changes

 -- Michael Vogt <michael.vogt@ubuntu.com>  Fri, 13 Nov 2015 10:21:18 +0100

goget-ubuntu-touch (0.31-0ubuntu1) wily; urgency=medium

  * Discriminate legacy from non legacy grub setup required installs.

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Wed, 16 Sep 2015 08:10:39 -0300

goget-ubuntu-touch (0.30-0ubuntu4) wily; urgency=medium

  * debian/control: Adding golang-uboot-go-dev as a build dep. 

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Tue, 15 Sep 2015 21:44:26 -0300

goget-ubuntu-touch (0.30-0ubuntu3) wily; urgency=medium

  * No change rebuild.

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Tue, 15 Sep 2015 21:36:27 -0300

goget-ubuntu-touch (0.30-0ubuntu2) wily; urgency=medium

  * No change rebuild. 

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Tue, 15 Sep 2015 21:22:34 -0300

goget-ubuntu-touch (0.30-0ubuntu1) wily; urgency=medium

  * Remove grub legacy handling for core. 

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Tue, 15 Sep 2015 16:12:21 -0300

goget-ubuntu-touch (0.29-0ubuntu1) wily; urgency=medium

  * Align core images to 512.
  * Copy original modprobe.d for core.

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Wed, 09 Sep 2015 17:44:05 -0300

goget-ubuntu-touch (0.28-0ubuntu1) wily; urgency=medium

  * Support setting the store id.
  * Support modprobe.d config hooks from ubuntu-core.

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Wed, 09 Sep 2015 13:59:52 -0300

goget-ubuntu-touch (0.27-0ubuntu1) wily; urgency=medium

  * Avoid double unmap errors (LP: #1472516)
  * Expose errors from external commands run in diskimage.

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Fri, 10 Jul 2015 14:28:26 -0300

goget-ubuntu-touch (0.26-0ubuntu1) wily; urgency=medium

  * snappy:
    - Avoid data races when downloading snappy assets.
    - Use common code for setting up the snappy bootloader.
    - Normalize kernel and initrd on deploy.
  * debian/control:
    - Add gettext dependency.
  * debian/changelog:
    - Remove extra whitespaces.

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Mon, 06 Jul 2015 11:00:45 -0300

goget-ubuntu-touch (0.25-0ubuntu1) wily; urgency=medium

  * Removing unneeded logic for creating /writable.
  * Add logic to prevent creating images lower than the unsupported size.
  * Add the snappy personal target.
  * Warn about the consequences of using --device-part for core and personal.

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Tue, 23 Jun 2015 14:09:54 -0300

goget-ubuntu-touch (0.24-0ubuntu2) wily; urgency=medium

  * Rebuild to use the latest ubuntu-snappy.

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Wed, 17 Jun 2015 16:54:55 -0300

goget-ubuntu-touch (0.24-0ubuntu1) wily; urgency=medium

  * core:
    - Stamp a yaml with install/setup information.
    - Remove code duplication and manage mounting/unmounting better.
  * Rebuild against latest ubuntu-snappy (LP: #1465879)

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Wed, 17 Jun 2015 10:04:06 -0300

goget-ubuntu-touch (0.23-0ubuntu2) wily; urgency=medium

  * No change rebuild for latest ubuntu-snappy

 -- Ricardo Salveti de Araujo <ricardo.salveti@canonical.com>  Wed, 10 Jun 2015 16:46:14 -0300

goget-ubuntu-touch (0.23-0ubuntu1) wily; urgency=medium

  * Update to us the latest snappy API.
  * Updating manpage

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Tue, 09 Jun 2015 10:57:15 -0300

goget-ubuntu-touch (0.22-0ubuntu2) wily; urgency=medium

  * debian/control: make ubuntu-snappy-cli a build-dep to side step the
    powerpc build failure from ubuntu-snappy.

 -- Sergio Schvezov <sergio.schvezov@canonical.com>  Mon, 08 Jun 2015 20:40:04 -0300

goget-ubuntu-touch (0.22-0ubuntu1) wily; urgency=medium

  [ Steve Langasek ]
  * Add UEFI support for grub.

  [ Sergio Schvezov ]
  * ubuntu-device-flash:
    - List channels for specific devices on query. (LP: #1460015)

  [ Jani Monoses ]
  * ubuntu-device-flash: support skipping tls on query.

  [ James Hunt ]

diskimage/common.go
//
// diskimage - handles ubuntu disk images
//
// Copyright (c) 2013 Canonical Ltd.
//
// Written by Sergio Schvezov <sergio.schvezov@canonical.com>
//
package diskimage

import (
	"bufio"
	"errors"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"syscall"

	"launchpad.net/goget-ubuntu-touch/sysutils"
)

// This program is free software: you can redistribute it and/or modify it
// under the terms of the GNU General Public License version 3, as published
// by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranties of
// MERCHANTABILITY, SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR
// PURPOSE.  See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program.  If not, see <http://www.gnu.org/licenses/>.

var debugPrint bool

func init() {
	if debug := os.Getenv("DEBUG_DISK"); debug != "" {
		debugPrint = true
	}
}

const (
	hardwareFileName = "hardware.yaml"
	kernelFileName   = "vmlinuz"
	initrdFileName   = "initrd.img"
)

const (
	partLayoutSystemAB = "system-AB"
)

var (
	syscallSync = syscall.Sync
)

type Image interface {
	Mount() error
	Unmount() error
	Format() error
	Partition() error
	BaseMount() string
}

type SystemImage interface {
	Boot() string
	System() string
	Writable() string
}

type CoreImage interface {
	Image
	SystemImage
	SetupBoot() error
	FlashExtra() error
}

type HardwareDescription struct {
	Kernel string `yaml:"kernel"`
	Dtbs   string `yaml:"dtbs"`
	Initrd string `yaml:"initrd"`
}

type BootAssetRawFiles struct {
	Path   string `yaml:"path"`
	Offset string `yaml:"offset"`
}

type BootAssetRawPartitions struct {
	Name string `yaml:"name"`
	Size string `yaml:"size"`
	Type string `yaml:"type"`
}

type BootAssetFiles struct {
	Path string `yaml:"path"`
	// Target is the deprecated target relative to $bootloader dir
	Target string `yaml:"target,omitempty"`
	// Dst is the destination relative to the actual boot partition
	Dst string `yaml:"dst,omitempty"`
}

type BootAssets struct {
	Files         []BootAssetFiles         `yaml:"files,omitempty"`
	RawFiles      []BootAssetRawFiles      `yaml:"raw-files,omitempty"`
	RawPartitions []BootAssetRawPartitions `yaml:"raw-partitions,omitempty"`
}

type OemDescription struct {
	Name    string `yaml:"name"`
	Version string `yaml:"version"`

	OEM struct {
		Hardware struct {
			Bootloader      string      `yaml:"bootloader"`
			PartitionLayout string      `yaml:"partition-layout"`
			Dtb             string      `yaml:"dtb,omitempty"`
			Platform        string      `yaml:"platform"`
			Architecture    string      `yaml:"architecture"`
			BootAssets      *BootAssets `yaml:"boot-assets,omitempty"`
		} `yaml:"hardware,omitempty"`

		Software struct {
			BuiltIn      []string `yaml:"built-in,omitempty"`
			Preinstalled []string `yaml:"preinstalled,omitempty"`
		} `yaml:"software,omitempty"`

		Store *struct {
			ID string `yaml:"id,omitempty"`
		}
	} `yaml:"oem,omitempty"`

	Config struct {
		UbuntuCore struct {
			Modprobe *string `yaml:"modprobe,omitempty"`
		} `yaml:"ubuntu-core,omitempty"`
	} `yaml:"config,omitempty"`

	rootDir string
}

func (o *OemDescription) SetRoot(rootDir string) {
	o.rootDir = rootDir
}

// SystemParts returns the system labels depending on the partition layout.
//
// The default is to return a flat structure for any unknown layout.
func (o *OemDescription) SystemParts() []string {
	switch o.OEM.Hardware.PartitionLayout {
	case partLayoutSystemAB:
		return []string{"a", "b"}
	default:
		return []string{""}
	}
}

func (o OemDescription) InstallPath() (string, error) {
	glob, err := filepath.Glob(fmt.Sprintf("%s/oem/%s/current", o.rootDir, o.Name))
	if err != nil {
		return "", err
	}

	if len(glob) != 1 {
		return "", errors.New("oem package not installed")
	}

	return glob[0], nil
}

func (o OemDescription) Architecture() string {
	return o.OEM.Hardware.Architecture
}

func (o *OemDescription) SetArchitecture(architecture string) {
	o.OEM.Hardware.Architecture = architecture
}

func (o OemDescription) PartitionLayout() string {
	return o.OEM.Hardware.PartitionLayout
}

func (o OemDescription) Platform() string {
	return o.OEM.Hardware.Platform
}

func (o *OemDescription) SetPlatform(platform string) {
	o.OEM.Hardware.Platform = platform
}

func sectorSize(dev string) (string, error) {
	out, err := exec.Command("blockdev", "--getss", dev).CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("unable to determine block size: %s", out)
	}

	return strings.TrimSpace(string(out)), err
}

// BaseImage implements the basic primitives to manage images.
type BaseImage struct {
	baseMount string
	hardware  HardwareDescription
	location  string
	oem       OemDescription
	parts     []partition
	partCount int
	size      int64
	rootSize  int
	label     string
}

// Mount mounts the image. This also maps the loop device.
func (img *BaseImage) Mount() error {
	if err := img.doMap(); err != nil {
		return err
	}

	baseMount, err := ioutil.TempDir(os.TempDir(), "diskimage")
	if err != nil {
		return err
	}

	//Remove Mountpoint if we fail along the way
	defer func() {
		if err != nil {
			if err := os.Remove(baseMount); err != nil {
				fmt.Println("WARNING: cannot remove", baseMount, "due to", err)
			}
		}
	}()

	// We change the mode so snappy can unpack as non root
	if err := os.Chmod(baseMount, 0755); err != nil {
		return err
	}

	for _, part := range img.parts {
		if part.fs == fsNone {
			continue
		}

		mountpoint := filepath.Join(baseMount, string(part.dir))
		if err := os.MkdirAll(mountpoint, 0755); err != nil {
			return err
		}

		dev := filepath.Join("/dev/mapper", part.loop)
		printOut("Mounting", dev, part.fs, "to", mountpoint)
		if out, errMount := exec.Command("mount", filepath.Join("/dev/mapper", part.loop), mountpoint).CombinedOutput(); errMount != nil {
			return ErrMount{dev: dev, mountpoint: mountpoint, fs: part.fs, out: out}
		}
		// this is cleanup in case one of the mounts fail
		defer func() {
			if err != nil {
				if err := exec.Command("umount", mountpoint).Run(); err != nil {
					fmt.Println("WARNING:", mountpoint, "could not be unmounted")
					return
				}

				if err := os.Remove(mountpoint); err != nil {
					fmt.Println("WARNING: could not remove ", mountpoint)
				}
			}
		}()
	}
	img.baseMount = baseMount

	return nil

}

// Unmount unmounts the image. This also unmaps the loop device.
func (img *BaseImage) Unmount() error {
	defer func() {
		if isMapped(img.parts) {
			fmt.Println("WARNING: could not unmap partitions")
		}
	}()

	if img.baseMount == "" {
		panic("No base mountpoint set")
	}

	syscallSync()

	for _, part := range img.parts {
		if part.fs == fsNone {
			continue
		}

		mountpoint := filepath.Join(img.baseMount, string(part.dir))
		if out, err := exec.Command("umount", mountpoint).CombinedOutput(); err != nil {
			lsof, _ := exec.Command("lsof", "-w", mountpoint).CombinedOutput()
			printOut(string(lsof))
			dev := filepath.Join("/dev/mapper", part.loop)
			return ErrMount{dev: dev, mountpoint: mountpoint, fs: part.fs, out: out}
		}
	}

	if err := os.RemoveAll(img.baseMount); err != nil {
		return err
	}
	img.baseMount = ""

	return img.doUnmap()
}

// doMap maps the image to loop devices
func (img *BaseImage) doMap() error {
	if isMapped(img.parts) {
		panic("cannot double map partitions")
	}

	kpartxCmd := exec.Command("kpartx", "-avs", img.location)
	stdout, err := kpartxCmd.StdoutPipe()
	if err != nil {
		return err
	}

	if err := kpartxCmd.Start(); err != nil {
		return err
	}

	loops := make([]string, 0, img.partCount)
	scanner := bufio.NewScanner(stdout)
	for scanner.Scan() {
		fields := strings.Fields(scanner.Text())

		if len(fields) > 2 {
			loops = append(loops, fields[2])
		} else {
			return fmt.Errorf("issues while determining drive mappings (%q)", fields)
		}
	}
	if err := scanner.Err(); err != nil {
		return err
	}

	if len(loops) != img.partCount {
		return ErrMapCount{expectedParts: img.partCount, foundParts: len(loops)}
	}

	mapPartitions(img.parts, loops)

	if err := kpartxCmd.Wait(); err != nil {
		return err
	}

	return nil
}

// doUnmap destroys loop devices for the partitions
func (img *BaseImage) doUnmap() error {
	if img.baseMount != "" {
		panic("cannot unmap mounted partitions")
	}

	for _, part := range img.parts {
		dmsetupCmd := []string{"dmsetup", "clear", part.loop}
		if out, err := exec.Command(dmsetupCmd[0], dmsetupCmd[1:]...).CombinedOutput(); err != nil {
			return &ErrExec{command: dmsetupCmd, output: out}
		}
	}

	kpartxCmd := []string{"kpartx", "-ds", img.location}
	if out, err := exec.Command(kpartxCmd[0], kpartxCmd[1:]...).CombinedOutput(); err != nil {
		return &ErrExec{command: kpartxCmd, output: out}
	}

	unmapPartitions(img.parts)

	return nil
}

// Format formats the image following the partition types and labels them
// accordingly.
func (img BaseImage) Format() (err error) {
	if err := img.doMap(); err != nil {
		return err
	}
	defer func() {
		if errUnmap := img.doUnmap(); errUnmap != nil {
			if err == nil {
				err = errUnmap
			} else {
				fmt.Println("WARNING: could not unmap partitions after error:", errUnmap)
			}
		}
	}()

	for _, part := range img.parts {
		dev := filepath.Join("/dev/mapper", part.loop)

		if part.fs == fsFat32 {
			cmd := []string{"mkfs.vfat", "-F", "32", "-n", string(part.label)}

			size, err := sectorSize(dev)
			if err != nil {
				return err
			}

			if size != "512" {
				cmd = append(cmd, "-s", "1")
			}

			cmd = append(cmd, "-S", size, dev)

			if out, err := exec.Command(cmd[0], cmd[1:]...).CombinedOutput(); err != nil {
				return &ErrExec{command: cmd, output: out}
			}
		} else {
			cmd := []string{"mkfs.ext4", "-F", "-L", string(part.label), dev}
			if out, err := exec.Command(cmd[0], cmd[1:]...).CombinedOutput(); err != nil {
				return &ErrExec{command: cmd, output: out}
			}
		}
	}

	return nil
}

// User returns the writable path
func (img BaseImage) Writable() string {
	if img.parts == nil {
		panic("img is not setup with partitions")
	}

	if img.baseMount == "" {
		panic("img not mounted")
	}

	return filepath.Join(img.baseMount, string(writableDir))
}

func (img BaseImage) pathToMount(dir directory) string {
	if img.parts == nil {
		panic("img is not setup with partitions")
	}

	if img.baseMount == "" {
		panic("img not mounted")
	}

	return filepath.Join(img.baseMount, string(dir))
}

//System returns the system path
func (img BaseImage) System() string {
	return img.pathToMount(systemADir)
}

// Boot returns the system-boot path
func (img BaseImage) Boot() string {
	return img.pathToMount(bootDir)
}

// BaseMount returns the base directory used to mount the image partitions.
func (img BaseImage) BaseMount() string {
	if img.baseMount == "" {
		panic("image needs to be mounted")
	}

	return img.baseMount
}

func (img *BaseImage) GenericBootSetup(bootPath string) error {
	// origins
	hardwareYamlPath := filepath.Join(img.baseMount, hardwareFileName)
	kernelPath := filepath.Join(img.baseMount, img.hardware.Kernel)
	initrdPath := filepath.Join(img.baseMount, img.hardware.Initrd)

	// populate both A/B
	for _, part := range img.oem.SystemParts() {
		path := filepath.Join(bootPath, part)

		printOut("Setting up", path)

		if err := os.MkdirAll(path, 0755); err != nil {
			return err
		}

		if err := sysutils.CopyFile(hardwareYamlPath, filepath.Join(path, hardwareFileName)); err != nil {
			return err
		}

		if err := sysutils.CopyFile(kernelPath, filepath.Join(path, kernelFileName)); err != nil {
			return err
		}

		if err := sysutils.CopyFile(initrdPath, filepath.Join(path, initrdFileName)); err != nil {
			return err
		}
	}

	oemRoot, err := img.oem.InstallPath()
	if err != nil {
		return err
	}

	return setupBootAssetFiles(img.Boot(), bootPath, oemRoot, img.oem.OEM.Hardware.BootAssets.Files)
}

func (img *BaseImage) FlashExtra() error {
	oemRoot, err := img.oem.InstallPath()
	if err != nil {
		return err
	}

	if bootAssets := img.oem.OEM.Hardware.BootAssets; bootAssets != nil {
		if bootAssets.RawPartitions != nil {
			if err := setupBootAssetRawPartitions(img.location, img.partCount, bootAssets.RawPartitions); err != nil {
				return err
			}
		}
		return setupBootAssetRawFiles(img.location, oemRoot, bootAssets.RawFiles)
	}

	return nil
}

func printOut(args ...interface{}) {
	if debugPrint {
		fmt.Println(args...)
	}
}
